### PR TITLE
[WIP] expired ballot checker

### LINKED ISSUE #21: Expired 'VotingResult' in 'VotingBox' must be moved to 'ReservedBox'
If `VotingResult` is stayed too long(almost over 2 minutes) in `VotingBox`, it should be moved to `ReservedBox`

### PR DESCRIPTION
**Pull request by sconeman**
### Github Issue

resolves #21 #22 #23 

### Possible Drawbacks

**STILL IN PROGRESS**
> Please check the assignee comment in file changed

Triple programming with @Charleslee522 @enessech @leejames00 @anarcher

### CODE REVIEW

**Comment by outsideris** [comment]
Please sign CLA @astonshnam

File: lib/ballot.go [hunk 1]
```diff
@@ -370,6 +370,22 @@ func (b *BallotBoxes) RemoveVotingResult(vr *VotingResult) (err error) {
...
+		err = sebakerror.ErrorVotingResultNotFound
+		return
+	}
+
+	b.Lock()
+	defer b.Unlock()
+
+	delete(b.Results, hash)
+	delete(b.Messages, hash)
+	// need to delete b.Sources
```
Comment (enessech) [inline lib/ballot.go]: @Charleslee522 @anarcher I want to create a method that uses `MessageHash` to clear the corresponding `VotingResult`. But it seems that the `Source` for `MessageHash` should also be cleared, and I am wondering how to fix it.

Comment (Charleslee522) [reply lib/ballot.go]: @ enessech
Oh... b.Sources couldn’t be removed by only hash, it needs voting result.
I figure, lazy removing is better.

Comment (anarcher) [reply lib/ballot.go]: I am not sure that works. When we remove VoltingResult, Can we remove `Source`? 

```
if vr,ok := b.Results[hash] ; ok {
   delete(b.Sources,vr.Source()]
}
```

Comment (enessech) [reply lib/ballot.go]: @anarcher 
```
 func (b *BallotBoxes) RemoveVotingResult(vr *VotingResult) (err error) {
 	if !b.HasMessageByHash(vr.MessageHash) {
 		err = sebakerror.ErrorVotingResultNotFound
 		return
 	}
 
 	b.Lock()
 	defer b.Unlock()
 
 	delete(b.Results, vr.MessageHash)
 	delete(b.Messages, vr.MessageHash)
 	delete(b.Sources, vr.Source)
 
  	return
  }
```
Using the method(`RemoveVotingResult`) in 357th line, we can remove `VotingResult` and `Source`

Comment (enessech) [reply lib/ballot.go]: @anarcher I will try to use this method.

Comment (Charleslee522) [reply lib/ballot.go]: > When we remove VoltingResult, Can we remove Source?

@anarcher Yes right.

> Using the method(RemoveVotingResult) in 357th line, we can remove VotingResult and Source

@enessech I remember that we save only Message hash and VotingResult hash, but now we need to save VotingResult.Source() also.

File: lib/ballot_expirer.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	e.stop <- q
+	<-q
+}
+
+func (e *BallotBoxExpireRemover) removeVotingResults() {
+	for _, hash := range e.prevBoxHashes {
+		if e.srcBox.HasMessageByHash(hash) { // srcBox.HasHash(h)?
+			if vr, ok := e.boxes.Results[hash]; ok {
+				e.boxes.RemoveVotingResult(vr)
+			}
```
Comment (enessech) [inline lib/ballot_expirer.go]: @Charleslee522 @anarcher I fixed thanks to @anarcher  :)

**Comment by enessech** [comment]
Now I will complete the test file.

File: lib/ballot_expirer.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		}
+	}
+}
+
+func (e *BallotBoxExpireRemover) makePrevHashesFromSrcBox() {
+	e.prevBoxHashes = make(map[string]string)
+
+	for hash, _ := range e.srcBox.Hashes {
+		if vr, ok := e.votingResults[hash]; ok {
+			if byteArr, err := sebakcommon.MakeObjectHash(vr); err == nil {
```
Comment (enessech) [inline lib/ballot_expirer.go]: @spikeekips @Charleslee522 When I do unit test, I get the following error message in this section.
```
"rlp: type map[sebakcommon.BallotState]sebak.VotingResultBallots is not RLP-serializable"
```
Isn't `map` object RLP-serializable?

**Comment by sconeman** [comment]
I will close this PR because we use ballotboxes anymore.

**Pull request closed without merging**